### ISSUE TITLE
[Bug]: Backward-origin compute straggler is undetectable: DDP all-reduce equalizes per-rank backward wall time, run collapses to global COMPUTE_BOUND

### ISSUE THREAD

**Issue by Pendu** [traceopt-ai/traceml#252]
### What happened?

In a controlled 4x T4 single-machine DDP experiment (traceml-ai 0.3.4, PyTorch 2.11.0+cu130, ResNet-18, 300 steps), a 200 ms/step busy-wait was injected into backward on rank 1 only.
Expected: a rank-named `COMPUTE_STRAGGLER` (the kind exists in `step_time/api.py`).
Observed: a global `COMPUTE_BOUND` with no rank named: "Model compute took 346.4ms of a 362.1ms average step."

The positive control in the same experiment works.
A 150 ms/batch dataloader stall injected on rank 2 only is correctly named: `INPUT_STRAGGLER`, "Rank r2 input wait was 199.7ms vs median rank r0 at 10.4ms."

### Why the detection cannot fire

Per-rank backward wall times from the run's `final_summary.json`:

| rank | backward_ms |
|---|---|
| 0 | 293.0 |
| 1 (injected +200 ms) | 292.2 |
| 2 | 293.0 |
| 3 | 293.1 |

The culprit rank is indistinguishable from the victims.
The DDP gradient all-reduce acts as a barrier inside backward: rank 1's backward is grad compute plus the injected delay, while every other rank's backward is grad compute plus an equal all-reduce wait.
Backward wall time equalizes exactly, so no per-rank comparison of current timings can recover the culprit.

The clean-straggler discount (`diagnostics/step_time/context.py`, `_clean_rank_timings`) can only explain a rank's backward delay through another rank's non-backward excess:

```
explained_bwd_delay_r = max(0, max(non_bwd) - non_bwd_r)
```

A backward-origin delay produces zero non-backward asymmetry, so `clean_step` stays symmetric, `_clean_straggler_evidence` returns `None`, and the run falls through to the global `COMPUTE_BOUND` rule.
Given its inputs the rule behaves correctly; the inputs cannot carry the signal.

Cross-check with an independent tool: Holistic Trace Analysis over per-rank torch.profiler traces of the same workload localizes the straggler immediately (rank 1 at 95% compute share of its timeline vs roughly 46% on the other ranks, which sit above 50% in NCCL wait).

### Scope

The blind spot is specific to delays that originate inside backward.
Input, H2D, forward, and optimizer-origin delays create non-backward asymmetry and are caught by the existing discount (see the `INPUT_STRAGGLER` control above).
A real compute straggler (thermal throttling, a slow or contended GPU) slows every phase, but backward is usually the largest compute phase, so most of the excess lands exactly in the blind spot.

### Relationship to open work

- #224 provides exactly the missing signal. Its motivation already notes that communication cost is "hidden inside backward/forward wall time or residual timing". With per-step communication wait measurable per rank, the victims' backward inflation becomes explainable (large comm wait) while the culprit's does not (excess compute, near-zero comm wait), which would let the existing `CleanStragglerRule` name the rank. Suggest adding a follow-up or acceptance criterion to #224: Step Time straggler detection consumes the communication signal to discount backward wait.
- #235 / #238 refine scoring and cause naming for stragglers the current visible-phase model can already detect. They keep that model, so on their own they cannot surface this case.

### Reproduce

4-rank DDP on one machine, any small CNN.
On rank 1 only, add a ~200 ms busy-wait inside backward (for example via a gradient hook on a late bucket).
Run under `traceml run` for ~300 steps.
Observed: global `COMPUTE_BOUND`, symmetric per-rank `backward_ms`, no rank named.
Expected: `COMPUTE_STRAGGLER` naming rank 1.

**Comment by Pendu** [comment]
Version scoping for this report:

- The experiment ran the released `traceml-ai` 0.3.4 (per the run's env manifest). `main` is equivalent for this purpose: `git diff v0.3.4..main -- src/traceml_ai/diagnostics/step_time/` is empty, so the behavior above is what current main ships.
- `version_0.3.5` does not fix this case, although it replaces the code cited above. The culprit-first model from #228 (`_build_rank_straggler_evidence` / `_rank_straggler_pair` in `diagnostics/step_time/context.py`) takes backward as the visible sync phase, picks the rank with the smallest backward as culprit, and gates on `victim_backward - culprit_backward` over iteration time. A backward-origin delay inflates the culprit's backward (excess compute) and every victim's backward (all-reduce wait) by the same amount, so the visible values stay symmetric by construction. Tracing the measured timings from this experiment through that path: backward is 292.2 to 293.1 ms across ranks, pair cost is about 0.8 ms, score is about 0.2% of a 362 ms iteration, far below the warn threshold, so the evidence builder returns `None` and the run still collapses to a global `COMPUTE_BOUND`. In addition, `COMPUTE_STRAGGLER` attribution on 0.3.5 is forward-excess only, and the FSDP branch comment in that function already states the constraint: "without explicit collective timing this rule should not emit compute stragglers from forward excess".
- Caveat: the 0.3.5 statement is a desk-check of the new code path against the 0.3.4-measured per-rank timings, not a rerun. The backward symmetry is created by the all-reduce barrier rather than by any version's instrumentation, so the result should carry over; happy to rerun the injected-straggler config on `version_0.3.5` to confirm empirically.

Conclusion unchanged: no per-rank comparison of backward wall time can separate excess compute from induced wait; the detector needs the per-step communication signal from #224.

**Comment by abhinavsriva** [comment]
@Pendu Thanks for the detailed reproduction. This behavior is expected with the telemetry TraceML currently records.

TraceML measures the outer backward wall time which includes both grad computation and DDP gradient sync. When a delay is inside backward other ranks absorb the delay as collective wait. Their backward wall times may therefore converge leaving no  per-rank signal.

The samecan occur under real conditions such as asymmetric network or NCCL slowdown. From the current Step Time measurements alone, TraceML cannot distinguish:

- a rank performing slower backward computation; from
- other ranks waiting longer in distributed collectives.

Therefore this is a known false negative. The global informational `COMPUTE_BOUND` result may still describe the backward envelope, but it cannot identify the cause or rank reliably.

We should document this limitation and add this reproduction as an acceptance case for #224. Until collective timing is available, TraceML should abstain from naming a compute-straggler rank in this scenario.
